kernel/proc.c

```diff
@@ -109,34 +120,57 @@ allocpid()
 static struct proc*
 allocproc(void)
 {
-  struct proc *p;
+  struct proc_list *cur_list;
+  struct proc* p;
+
+  cur_list = bd_malloc(sizeof(struct proc_list));
+  p = bd_malloc(sizeof(struct proc));
+  cur_list->cur_proc = p;
+  // No memory for new proc
+  if (p == 0 || cur_list == 0) {
+    return 0;
+  }
+  initlock(&p->lock, "proc");
+  acquire(&p->lock);
+  p->pid = allocpid();
 
-  for(p = proc; p < &proc[NPROC]; p++) {
-    acquire(&p->lock);
-    if(p->state == UNUSED) {
-      goto found;
-    } else {
-      release(&p->lock);
+  // Use prealloced kstacks if possible
+  uint64 va;
+  int found_va = 0;
+  acquire(&kstacks_lock);
+  for (int idx = 0; idx < NPROC_KSTACK; idx++) {
+    if (free_kstacks[idx]) {
+      va = kstacks[idx];
+      free_kstacks[idx] = 0;
+      found_va = 1;
+      break;
     }
   }
-  return 0;
+  release(&kstacks_lock);
+  if (!found_va) {
+    char *pa = kalloc();
+    if(pa == 0)
+      panic("kalloc");
 
-found:
-  p->pid = allocpid();
+    va = KSTACK(p->pid + NPROC_KSTACK);
```

Comment:
It will overflow in any implementation as far as I understand. So the problem is - we need to allocate the KSTACK for each process and the only way to do it and not lose memory EVERY time is to keep some array of possibilities. But of course as you  pointed out there can be times when all preallocated kstacks are taken so we have no choice but to allocate more memory (maybe as a possibility we can kill a proccess since it is widely used in lab3 but for now let's pretend we can't). So sure we will eventually overflow but so will any other implementation. This formula is chosen by the reason actually mentioned by you ) since kstacks from 0 to NPROC_KSTACK and we need to ensure our new kstack won't interfere with them we can user pid + NPROC_KSTACK. Since it is unique for each process and strictly greater then NPROC_KSTACK.